src/mam/mod_mam_rdbms_arch_async.erl

```diff
@@ -1,80 +1,60 @@
+-ignore_xref([archive_pm_message/3, mam_archive_sync/2]).
 
--export([archive_pm_message/3, archive_muc_message/3]).
--export([mam_archive_sync/2, mam_muc_archive_sync/2]).
--export([flush_pm/2, flush_muc/2]).
-
--type writer_type() :: pm | muc.
-
--ignore_xref([archive_pm_message/3, archive_muc_message/3]).
--ignore_xref([mam_archive_sync/2, mam_muc_archive_sync/2]).
+-export([make_pool_opts/2, prepare_insert_queries/2]).
 
--spec archive_pm_message(_Result, mongooseim:host_type(), mod_mam:archive_message_params()) ->
-          ok | {error, timeout}.
+-spec archive_pm_message(_Result, mongooseim:host_type(), mod_mam:archive_message_params()) -> ok.
 archive_pm_message(_Result, HostType, Params = #{archive_id := ArcID}) ->
-    PoolName = mongoose_async_pools:pool_name(HostType, pm_mam),
-    wpool:cast(PoolName, {task, Params}, {hash_worker, ArcID}).
-
--spec archive_muc_message(_Result, mongooseim:host_type(), mod_mam:archive_message_params()) ->
-    ok.
-archive_muc_message(_Result, HostType, Params0 = #{archive_id := RoomID}) ->
-    Params = mod_mam_muc_rdbms_arch:extend_params_with_sender_id(HostType, Params0),
-    PoolName = mongoose_async_pools:pool_name(HostType, muc_mam),
-    wpool:cast(PoolName, {task, Params}, {hash_worker, RoomID}).
+    mongoose_async_pools:put_task(HostType, pm_mam, ArcID, Params).
 
 -spec mam_archive_sync(term(), mongooseim:host_type()) -> term().
 mam_archive_sync(Result, HostType) ->
     mongoose_async_pools:sync(HostType, pm_mam),
     Result.
 
--spec mam_muc_archive_sync(term(), mongooseim:host_type()) -> term().
-mam_muc_archive_sync(Result, HostType) ->
-    mongoose_async_pools:sync(HostType, muc_mam),
-    Result.
-
 %%% gen_mod callbacks
 -spec start(mongooseim:host_type(), gen_mod:module_opts()) -> any().
 start(HostType, Opts) ->
-    [ start_pool(HostType, Mod) || Mod <- maps:to_list(Opts) ].
+    {PoolOpts, Extra} = make_pool_opts(pm, Opts),
+    prepare_insert_queries(pm, Extra),
+    mongoose_metrics:ensure_metric(HostType, ?PER_MESSAGE_FLUSH_TIME, histogram),
+    mongoose_metrics:ensure_metric(HostType, ?FLUSH_TIME, histogram),
+    ejabberd_hooks:add(mam_archive_sync, HostType, ?MODULE, mam_archive_sync, 50),
+    ejabberd_hooks:add(mam_archive_message, HostType, ?MODULE, archive_pm_message, 50),
+    mongoose_async_pools:start_pool(HostType, pm_mam, PoolOpts).
 
 -spec stop(mongooseim:host_type()) -> any().
 stop(HostType) ->
-    Opts = gen_mod:get_loaded_module_opts(HostType, ?MODULE),
-    [ stop_pool(HostType, Mod) || Mod <- maps:to_list(Opts) ].
+    ejabberd_hooks:delete(mam_archive_message, HostType, ?MODULE, archive_pm_message, 50),
+    ejabberd_hooks:delete(mam_archive_sync, HostType, ?MODULE, mam_archive_sync, 50),
+    mongoose_async_pools:stop_pool(HostType, pm_mam).
 
 -spec config_spec() -> mongoose_config_spec:config_section().
 config_spec() ->
-    mongoose_async_pools:config_spec().
+    mod_mam_meta:async_config_items().
```

Comment:
I've just realized that these function are completely unused as this module is not configurable directly - and not intended to be. IMO we could just remove this with its muc equivalent. What do you think?